### PR TITLE
feat: generate Dockerfile on-the-fly

### PR DESCRIPTION
**Pull request by Krzmbrzl**
This aims at implementing #63

### CODE REVIEW

File: scripts/runDockerFile.sh
Comment (Krzmbrzl) [inline scripts/runDockerFile.sh:3]: The used shell-linter complains about `$@` potentially being re-split - but this is actually exactly what we want in this case, since we want to pass the parameters separately and not as one big, concatenated chunk

File: scripts/buildDockerfile.sh
Comment (Krzmbrzl) [inline scripts/buildDockerfile.sh:6]: ShellCheck complains about the unquoted expansion of this variable. They [recommend](https://github.com/koalaman/shellcheck/wiki/SC2086) using arrays instead, but since we are using environment variables for this purpose here, that [doesn't work](https://unix.stackexchange.com/a/393096/203826).

Thus, I don't see a way to fix this

Comment (jidicula) [reply scripts/buildDockerfile.sh:6]: could we...

* hardcode the `-v` flag since that doesn't change, and quote the variable expansion for the workspace variable passed in from `action.yml`?
* directly reference the workspace variable in the script? you might have to use the format `$GITHUB_WORKSPACE` instead of the `${}` format: 
```
docker build . -v "$GITHUB_WORKSPACE":/clang-format-action/workspace --file Dockerfile --tag local/clang-format-action:latest
```

I noticed you pushing a lot of commits while working on the draft PR, so it's entirely possible you tried these - let me know if you've already exhausted those options 😅

**Review by jidicula — CHANGES_REQUESTED** [review]
I just ran a quick check to see the runtime difference with this refactor:

[before](https://github.com/jidicula/test-clang-format-action/runs/4695542970?check_suite_focus=true): 13 s

[this PR](https://github.com/jidicula/test-clang-format-action/runs/4695584916?check_suite_focus=true): 71 s

This is on my old testing repo at https://github.com/jidicula/test-clang-format-action . The files are pretty much negligible in size compared to most codebases using this action, so we're pretty close to polling the setup time before clang-format actually runs (I _think_ a more minimal example would be similar in runtime since clang-format is pretty fast to begin with).

As it stands, I can't accept this refactor with such a drastic slowdown in the tiniest of C codebases. I haven't looked closely to see where the slowdown is being introduced, but I'd hazard a guess that it's something to do with running a Docker instance inside a composite action. I'll leave it up to you if you want to keep hacking at this PR to improve it, or if you want to abandon it - I know you spent a lot of time on it already, so no hard feelings if you'd prefer to not spend any more cycles on this.

**Comment by Krzmbrzl** [comment]
I would assume the time is mostly spent on downloading the LLVM image. From my tests here on this PR, it always seemed as if GitHub would cache this image though and thus the successive runs were significantly faster than the initial one.

That being said though, the action taking about a minute is more than worth it, if in exchange it doesn't break just because some Linux distro goes EOL :shrug: 
And at least in our context, we not only run the clang-format action, but also CI runs that actually build and test our code. Compared to that the time required for the clang-format action is completely insignificant.

File: scripts/buildDockerfile.sh
Comment (Krzmbrzl) [reply scripts/buildDockerfile.sh:6]: yeah, I think that should be possible. It removes the possibility of specifying any other option though. However, this doesn't seem to be a problem as of now...

**Comment by Krzmbrzl** [comment]
Okay so I just did a few tests myself (https://github.com/Krzmbrzl/test-clang-format-action/actions) and repeatedly executing the action doesn't seem to help.
The majority of the time spent in the action is spent on building the docker image, which I assume is happening somewhere behind the scenes when choosing a Docker-action in the first place. If that's true, then I don't think there is much we/I can do about it, except maybe trying to use some caching :eyes:

**Comment by jidicula** [comment]
I wonder if we can get a speedup if I mirror silkeh's LLVM images repository but using the GitHub container registry instead of Docker's.

**Comment by Krzmbrzl** [comment]
If that is not too much effort, I guess it would be worth a try :thinking:

**Comment by jidicula** [comment]
> If that is not too much effort, I guess it would be worth a try 🤔

I'll try to get it set up this week/end.

**Comment by jidicula** [comment]
I've got GitHub Packages working for setting up a container registry [here](https://github.com/jidicula/clang-format-action/pkgs/container/clang-format/versions). However, I'm trying to take a different approach than yours in #75 to see if it's possible to just call the script from inside a composite action, and modifying the script to use a `docker run` command on one of my `clang-format` images. Not sure if it'll work though, there seem to be some difficulties with GitHub Actions not using a TTY...

The download speed from ghcr.io does seem to be dramatically faster than from Dockerhub, though!

**Comment by jidicula** [comment]
Update: I've realized those images in my ghcr repo are missing some DLLs, so they're not functional at the moment.

**Comment by jidicula** [comment]
Ok, the images [here](https://github.com/jidicula/clang-format-action/pkgs/container/clang-format/versions) seem to be correctly constructed now. You can try them out locally with `docker run --rm -it -v "$(pwd)":/workdir -w /workdir ghcr.io/jidicula/clang-format:"13" -n --Werror --style=file --fallback-style=LLVM addition.c`

**Comment by jidicula** [comment]
[Victory](https://github.com/jidicula/clang-format-action/actions/runs/1700616361)! I'll finish cleaning up that PR but for now it looks like this PR is no longer needed.  I'll keep it open for now, and close this once #75 is merged.

**Comment by Krzmbrzl** [comment]
Great!

If I parse your PR correctly, the main difference is still only that you are now using your mirrored Docker images, right? Very interesting that this makes such a huge difference :eyes:

**Comment by jidicula** [comment]
I think it's because the hosting for ghcr is within the same cloud as GitHub Actions, so the image download speed bottleneck is much lower, compared to fetching from Dockerhub 🤔

**Comment by Krzmbrzl** [comment]
Yeah that sounds like a reasonable explanation. Anyway, great that it is working now! :+1:

**Pull request closed without merging**
